blog/2022/05-10-resoto-graph-traversal/index.md

```diff
@@ -0,0 +1,119 @@
+
+Since this droplet is attached to a load balancer, we want to find out the public IP address. If multiple load balancers were attached to the same droplet, we would get the list of all IP addresses.
+
+```bash
+> search is(digitalocean_droplet) and id=289061882 <-- is(load_balancer) |
+  jq .public_ip_address
+// highlight-start
+​137.121.131.102
+// highlight-end
+```
+
+Resoto allows for walking not only one step but multiple steps. You can define a specific number of steps to walk and a range of steps. You can even specify the range as unlimited, which would traverse the graph to the root or leave, depending on the direction of the walk. The following query will select all DigitalOcean VPCs and then walk the graph outbound, returning all nodes that can be reached starting from the VPC up to two steps. This list would never include snapshots since snapshots can not be reached in two steps starting from the VPC.
+
+```bash
+> search is(digitalocean_vpc) -[0:2]-> | count kind
+// highlight-start
+​digitalocean_app: 1
+​digitalocean_database: 1
+​digitalocean_load_balancer: 2
+​digitalocean_kubernetes_cluster: 2
+​digitalocean_vpc: 2
+​digitalocean_volume: 6
+​digitalocean_droplet: 6
+​total matched: 20
+​total unmatched: 0
+// highlight-end
+```
+
+If we can walk outbound with `-->` and inbound with `<--`, there has to be an option to walk in both directions. The combination of the two approaches is expressed with `<-->`. It is possible to get the list of all reachable resources starting from a specific node. We use the above droplet and ask for all attached resources.
+
+```bash
+> search is(digitalocean_droplet) and id=289061882 <-[0:]-> | count kind
+// highlight-start
+​digitalocean_droplet: 1
+​digitalocean_load_balancer: 1
+​digitalocean_kubernetes_cluster: 1
+​digitalocean_region: 1
+​digitalocean_vpc: 1
+​graph_root: 1
+​cloud: 1
+​digitalocean_image: 1
+​digitalocean_team: 1
+​digitalocean_project: 1
+​digitalocean_volume: 2
+​total matched: 12
+​total unmatched: 0
+// highlight-end
+```
+
+The traversal definition `<-[0:]->` might look strange at first glance. It defines to walk inbound and outbound starting from the current node (`0`). Since the end of the walk is not specified, it will walk until there is no more node to traverse. It will walk to the graph root inbound and all possible leaves outbound. The result is counted by kind. To make this information even more helpful, we can create a graph diagram out of this query using [graphviz](https://www.graphviz.org/) and the [dot format](/docs/reference/cli/format). The same query as above is used while edges get included in the result.
```

Comment:
```suggestion
The traversal definition `<-[0:]->` might look strange at first glance. It defines to walk inbound and outbound starting from the current node (`0`). Since the end of the walk is not specified, it will walk until there are no more nodes to traverse. It will walk to the graph root inbound and all possible leafs outbound. The result is then counted by kind. To make this information even more helpful, we can create a graph diagram out of this search using [Graphviz](https://www.graphviz.org/) and the [dot format](/docs/reference/cli/format). The same search as above is used with edges included in the result.
```